### ISSUE TITLE
BERT block transfers could be easier to implement if reply will contain last _received_ sub-block

### ISSUE THREAD

**Issue by smolboarm** [core-wg/coap-tcp-tls#183]
Currently sender of BERT blocks with PUT or POST requests should persist transfer state (full transfer payload position) between blocks, because state can't be recovered from ACK message. With usual block transfers position can be recovered from ACK and sender does not need to persist position locally.

This is caused because receiver replies with same BlockOption as it received. This BlockOption contains **beginning** 1k block number.

Proposal: receiver should reply in ACK 231_CONTINUE **last received** 1k block number, instead of just copying block option. Then sender can position to next block just incrementing block number from ACK, like it is done for non-BERT block transfers.

Current:

CoAP Client                             CoAP Server
     |                                             |
     | PUT, /options, 1:0/1/BERT(8192)     ------> |   -- here client should save that it sent 8 sub-blocks
     |                                             |
     | <------   2.31 Continue, 1:**0**/1/BERT         |      -- adjust position by saved 8 sub-blocks
     |                                             |
     | PUT, /options, 1:8/1/BERT(16384)    ------> |  -- save 16 sub-blocks
     |                                             |
     | <------   2.31 Continue, 1:**8**/1/BERT         |
     |                                             |
     | PUT, /options, 1:24/0/BERT(5683)    ------> |
     |                                             |
     | <------   2.04 Changed, 1:**24**/0/BERT         |
     |                                             |

Proposed:

CoAP Client                             CoAP Server
     |                                             |
     | PUT, /options, 1:0/1/BERT(8192)     ------> |  
     |                                             |
     | <------   2.31 Continue, 1:**7**/1/BERT         |    -- just increment by 1
     |                                             |
     | PUT, /options, 1:8/1/BERT(16384)    ------> |     -- and we have position of next block to send
     |                                             |
     | <------   2.31 Continue, 1:**23**/1/BERT         |   -- again just increment by 1
     |                                             |
     | PUT, /options, 1:24/0/BERT(5683)    ------> |  
     |                                             |
     | <------   2.04 Changed, 1:**27**/0/BERT         |
     |                                             |

In this case logic and code for both Block1 and Block2 options internally will be same. And more over it will be same with non-BERT block transfers. Data receiver anyway has to count number of 1k sub-blocks received to request next blocks (GET request) or to reply ACK (PUT/POST in proposed solution).

**Comment by cabo** [comment]
I don't think this is an improvement.  Right now, an invariant of the descriptive usage of the block number is that (times the block size, or 1024 for szx==7) it indicates where the response payload fits in the overall body.  With your change, it would be necessary to compute the position from the size of the actual BERT block to get that information.  This also gets weird for the last block.  I don't think computing the increment from the next request by adding `payload_size >> 10` for BERT is particularly onerous (and it happens to work both for szx == 6 and szx == 7).

**Comment by cabo** [comment]
So to clarify, you change `new_block_to_request = block_returned + 1` into `new_block_to_request = block_returned + (payload_size/block_size)`.  That is not a lot of work (and you can actually use a `<<` instead of a `/`.)

**Comment by smolboarm** [comment]
I worry only about additional state that should be stored undefined time (last sent payload size, or last sent position). While it can be easily recovered from ACK message with small changes.

The problem - is how long to store this position, if we have packets timeout. Especially in case of CoAP/TCP where no ACK expected by default. We can end up with storing transaction state for undefined time. Or transaction state storing time should be defined in the spec.

Other way - send back last received payload position or last received block payload size in additional header in ACK response.
We can't rely on Max-Message-Size to determine last sent payload, because this setting can be changed between sending PUT / POST block and receiving ACK.

**Comment by cabo** [comment]
Hi @smolboarm, maybe I don't understand the scenario.

I was coming from my block-wise client code, which essentially adds 16 (1 << 4) to the block option received in the response to build the next request option.  There is no long-term state involved.  When you change the 1 to `(received_payload_size/block_size)`, there is still no long-term state involved -- all this information is in the response the code is reacting to.  (With szx==7, there needs to be a `max(szx, 6)` in that shift.  That is a bit ugly, but unrelated to the question here.)
